gordon/main.py

```diff
@@ -83,20 +86,76 @@ def setup(config_root=''):
     return config
 
 
-def _log_or_exit_on_exceptions(errors, debug):
+def _log_or_exit_on_exceptions(base_msg, exc, debug):
     log_level_func = logging.warn
     if not debug:
         log_level_func = logging.error
 
-    base_msg = 'Plugin "{name}" was not loaded:'
-    for name, exc in errors:
-        msg = base_msg.format(name=name)
-        log_level_func(msg, exc_info=exc)
+    if isinstance(exc, list):
+        for exception in exc:
+            log_level_func(base_msg, exc_info=exception)
+    else:
+        log_level_func(base_msg, exc_info=exc)
 
     if not debug:
         raise SystemExit(1)
 
 
+def _assert_required_plugins(plugins, debug):
+    required_providers = {
+        'event_consumer': False,
+        'enricher': False,
+        'publisher': False,
+    }
+    for plugin in plugins:
+        if interfaces.IEventConsumerClient.providedBy(plugin):
+            required_providers['event_consumer'] = True
+        elif interfaces.IEnricherClient.providedBy(plugin):
+            required_providers['enricher'] = True
+        elif interfaces.IPublisherClient.providedBy(plugin):
+            required_providers['publisher'] = True
+
+    missing = []
+    msg = ('The provider for "{name}" interface is not configured for the '
+           'Gordon service or is not implemented.')
+    for provider, available in required_providers.items():
+        if not available:
+            exc = exceptions.MissingPluginError(msg.format(name=provider))
+            missing.append(exc)
+    if missing:
+        base_msg = 'Issue running plugins: '
+        _log_or_exit_on_exceptions(base_msg, missing, debug=debug)
+
+
+def _gather_runnable_plugins(plugins, debug):
+    plugins_to_run = []
+    for plugin in plugins:
+        if interfaces.IEventConsumerClient.providedBy(plugin):
+            # TODO (lynn): this should be switched out for adding
+            # the "verify interface implementation" ability. See
+            # https://docs.zope.org/zope.interface/verify.html
+            if not hasattr(plugin, 'run') or \
+                    not asyncio.iscoroutinefunction(plugin.run):
+                msg = (f'Implemention "{plugin}" of the required '
+                       '"IEventConsumerClient" interface does not have the '
+                       'necessary `run` method.')
+                exc = exceptions.InvalidPluginError(msg)
+                _log_or_exit_on_exceptions(msg, exc, debug)
+                continue
+            plugins_to_run.append(plugin)
+        elif hasattr(plugin, 'run'):
```

Comment:
Should this also test for async?

Reply:
Maybe not _test_, but yea I think adding a condition of `asyncio.iscoroutinefunction` is reasonable.

Reply:
Actually I do already have this check [here](https://github.com/spotify/gordon/pull/17/commits/d427d4dfe735b5c79590fb32af68879a7ea5495c#diff-6ef2c41f2a81ebf1e2f3f7c8065b9160R138).

Reply:
That's for plugins that implement an interface.  What about other runnables?  Can you do an await on them in run() if they aren't async?

Reply:
ah right missed that part. I was like, I swore I already typed `asyncio.iscoroutinefunction` somewhere here..

Reply:
Can you add a test for this case?  The existing one only hits the interface cases.